### PR TITLE
[luci/pass] Introduce RemoveUnnecessaryStridedSlicePass

### PR DESCRIPTION
**Pull request by a-zh**
This commit adds pass for unnecessary StridedSlice operator

Draft: https://github.com/Samsung/ONE/pull/5876

Signed-off-by: Anton Zhukov <anton.v.zhukov@gmail.com>

### CODE REVIEW

**Review by seanshpark — CHANGES_REQUESTED** [review]
Please split PRs that have different `compiler/modules`

**Comment by seanshpark** [comment]
May I ask which issue is related to this Pass?

**Comment by a-zh** [comment]
> May I ask which issue is related to this Pass? Please ping me from that issue.

Please, find related issue https://github.com/Samsung/ONE/issues/5757

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 1]
```diff
@@ -43,34 +43,45 @@ bool remove_no_effect_strided_slice(luci::CircleNode *node)
   if (target_node == nullptr)
     return false;
 
-  auto begin_const = dynamic_cast<luci::CircleConst *>(target_node->begin());
+  auto begin_const = loco::must_cast<luci::CircleConst *>(target_node->begin());
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Why is this chage needed?

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 2]
```diff
@@ -43,34 +43,45 @@ bool remove_no_effect_strided_slice(luci::CircleNode *node)
...
 
+  auto end_const = loco::must_cast<luci::CircleConst *>(target_node->end());
+  if (end_const == nullptr)
+    return false;
+
   for (uint32_t i = 0; i < input_node->rank(); i++)
   {
-    if (value_from_circle_const(begin_const, i) != 0)
+    int64_t begin_value = value_from_circle_const(begin_const, i);
+    if (begin_value != 0)
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:65]: why is this change needed?

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+bool RemoveUnnecessaryStridedSlicePass::run(loco::Graph *g)
+{
+  bool changed = false;
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    auto circle_node = loco::must_cast<luci::CircleNode *>(node);
+    if (remove_no_effect_strided_slice(circle_node))
+    {
+      changed = true;
+    }
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:121]: It would be better to use like
```cxx
  if (target_node = dynamic_cast<luci::CircleStridedSlice *>(node))
  {
     if (remove_no_effect_strided_slice(target_node))
      changed = true;
  }
```
and remove the same check in `remove_no_effect_strided_slice()`

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  if (input_node == nullptr)
+    return false;
+
+  auto end_const = dynamic_cast<luci::CircleConst *>(target_node->end());
+  if (end_const == nullptr)
+    return false;
+
+  for (uint32_t i = 0; i < input_node->rank(); i++)
+  {
+    int64_t begin_value = value_from_circle_const(begin_const, i);
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:64]: `begin_value` is only used once to compare so it would be better not to declare,

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+ *          |
+ *    [CircleStridedSlice]
+ *          |
+ *    [CircleNode]
+ *
+ * AFTER
+ *
+ *    [CircleNode]
+ *          |
+ *    [CircleNode]
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:106]: `[CircleStridedSlice]` is not REMOVED. actually it's disconnected after this pass.
```
 *          |
 *    [CircleNode]
 *          |
 *    [CircleNode]   [CircleStridedSlice]
 *          |
```

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  // Output connect
+  auto output = g->nodes()->create<luci::CircleOutput>();
+  output->from(strided_slice);
+  auto graph_output = g->outputs()->create();
+  output->index(graph_output->index());
+}
+
+} // namespace
+
+TEST(RemoveUnnecessaryStridedSlicePass, remove_unnecessary_strided_slice)
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:85]: Can you use `basic_case` for `remove_unnecessary_strided_slice` ?

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    auto strided_slice = dynamic_cast<luci::CircleStridedSlice *>(node);
+    if (strided_slice == nullptr)
+      continue;
+    strided_slice_node = strided_slice;
+    break;
+  }
+  ASSERT_EQ(nullptr, strided_slice_node);
+}
+
+TEST(RemoveUnnecessaryStridedSlicePass, remove_unnecessary_strided_slice_NEG)
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:114]: can you use `basic_fail_case_NEG` ?

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp (hunk 6 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp (hunk 5 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp (hunk 4 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp (hunk 3 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  assert(node->dtype() == loco::DataType::S64 || node->dtype() == loco::DataType::S32);
+
+  if (node->dtype() == loco::DataType::S64)
+    return node->at<loco::DataType::S64>(idx);
+  return static_cast<int64_t>(node->at<loco::DataType::S32>(idx));
+}
+
+bool remove_no_effect_strided_slice(luci::CircleStridedSlice *target_node)
+{
+  auto begin_const = loco::must_cast<luci::CircleConst *>(target_node->begin());
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:42]: `must_cast` will throw if input node type is not the template type. you should use plain dynamic_cast

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * @brief   Return value in CircleConst.
+ * @details Return value in position on CircleConst with int64 format.
+ *          Begin must be larger than or equal to 0. Size must be larger
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:27]: Cannot find `Begin` and `Size` in this method.

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

**Comment by seanshpark** [comment]
@nnfw-bot test nncc-release

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Comment by struss** [comment]
Sorry for late reply. I tested on tensorflow v2.3 and have some question. while I create model like below, values and shapes are not matching with your conditions. 
Screenshot 2021-01-26 at 13 57 25
So, I think it is nice for this pass to add a `luci-pass-value-test` on it. 
for reference, https://github.com/Samsung/ONE/pull/5807

**Comment by a-zh** [comment]
> Sorry for late reply. I tested on tensorflow v2.3 and have some question. while I create model like below, values and shapes are not matching with your conditions.

I see the modified tensor form after the current `strided_slice` operator. The previous shape is (3,2,3) and new shape is (0,2,3). And we got necessary `strided_slice`. And because the pass was skipped.

> So, I think it is nice for this pass to add a `luci-pass-value-test` on your draft.
> for reference, #5807

I will try to add a test case.

**Comment by a-zh** [comment]
> > So, I think it is nice for this pass to add a `luci-pass-value-test` on your draft.
> > for reference, #5807
> 
> I will try to add a test case.

Please, find PRs:
https://github.com/Samsung/ONE/pull/5852
https://github.com/Samsung/ONE/pull/5853

**Comment by a-zh** [comment]
@struss @seanshpark **PTL**

**Review by mhs4670go — APPROVED** [review]
LGTM

**Comment by a-zh** [comment]
@struss @seanshpark Please take a look this PR.

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  {
+    strides->at<loco::DataType::S32>(i) = -1;
+  }
+
+  auto end = g->nodes()->create<luci::CircleConst>();
+  end->dtype(loco::DataType::S32);
+  end->size<loco::DataType::S32>(input_shape.size());
+  end->rank(1);
+  end->dim(0).set(input_shape.size());
+  end->shape(input_shape);
```
Comment (struss) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:66]: I think this need to set value on end(CircleConst) at this.

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:66]: We have `end->shape(input_shape);`
Which one of another initializer are needed?

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  // Strides create
+  auto strides = g->nodes()->create<luci::CircleConst>();
+  strides->dtype(loco::DataType::S32);
+  strides->size<loco::DataType::S32>(input_shape.size());
+  strides->rank(1);
+  strides->dim(0).set(input_shape.size());
+  for (int i = 0; i < input_shape.size(); ++i)
+  {
+    strides->at<loco::DataType::S32>(i) = -1;
+  }
```
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:59]:
```suggestion
  auto strides = g->nodes()->create<luci::CircleConst>();
  strides->dtype(loco::DataType::S32);
  strides->rank(1);
  strides->dim(0).set(input_shape.size());
  strides->size<loco::DataType::S32>(input_shape.size());
  for (int i = 0; i < input_shape.size(); ++i)
  {
    strides->at<loco::DataType::S32>(i) = -1;
  }
```

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp (hunk 10 above)
Comment (seanshpark) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp:66]:
```suggestion
  end->dtype(loco::DataType::S32);
  end->rank(1);
  end->dim(0).set(input_shape.size());
  end->size<loco::DataType::S32>(input_shape.size());
  for (int i = 0; i < input_shape.size(); ++i)
  {
    strides->at<loco::DataType::S32>(i) = ...;
  }
```
we don't need to set `end->shape(input_shape);` as `rank()` and `dim(..).set(..)` will set the shape.
you need to call ` strides->at<loco::DataType::S32>(i) = ...;` to set values of this const

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp]: Thanks a lot. Fixed.

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp (hunk 11 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.test.cpp]: Apply.

**Review by mhs4670go — DISMISSED** [review]
LGTM

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * AFTER
+ *
+ *    [CircleNode]
+ *          |
+ *    [CircleNode]   [CircleStridedSlice]
+ *
+ * StridedSlice OP has no effect if,
+ *    1. Static Shape : begin_const[idx] is 0 AND strides_const[idx] is (-1 OR input_dimension[idx])
+ *    2. Dynamic Shape : begin_const[idx] is 0 AND strides_const[idx] is -1
```
Comment (struss) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:100]: I think we need to add a comment about `end` value condition. follow from your logic above, end value is same with `(-1 OR input_dimension[idx])`

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+  auto input_node = loco::must_cast<luci::CircleNode *>(target_node->input());
+  for (uint32_t i = 0; i < input_node->rank(); i++)
+  {
+    if (value_from_circle_const(begin_const, i) != 0)
+      return false;
+
+    int64_t strides_value = value_from_circle_const(strides_const, i);
+    if (strides_value == -1)
+      continue;
+
```
Comment (struss) [inline compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp:61]: this checking logic seems not right for this. on [Tensorflow explain](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/ops/array_ops.py#L1129-L1133), Stride means 
```
  Roughly speaking, this op extracts a slice of size `(end-begin)/stride`
  from the given `input_` tensor. Starting at the location specified by `begin`
  the slice continues by adding `stride` to the index until all dimensions are
  not less than `end`.
  Note that a stride can be negative, which causes a reverse slice.
``` 
so, if `-1` in stride will make a result to reverse of input.

Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: OK. Done

File: compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp (hunk 12 above)
Comment (a-zh) [reply compiler/luci/pass/src/RemoveUnnecessaryStridedSlicePass.cpp]: Done

**Comment by a-zh** [comment]
@struss Please take a look PR

**Comment by struss** [comment]
Different from `Slice`, `StridedSlice` need to check option params. (like shrink_axis_mask).

**Comment by a-zh** [comment]
> Different from `Slice`, `StridedSlice` need to check option params. (like shrink_axis_mask).

I thought about it. But why? The sign for excluding the `StridedSlice` operator is only the equality of the input and output shape? And what test criteria is still required?

**Comment by struss** [comment]
> > Different from `Slice`, `StridedSlice` need to check option params. (like shrink_axis_mask).
> 
> I thought about it. But why? The sign for excluding the `StridedSlice` operator is only the equality of the input and output shape? And what test criteria is still required?

I just think this logic not consider the case with option. if we don't consider this, it could make result not matching with original. so, I think this point need to be updated before something goes wrong.

**Comment by a-zh** [comment]
@struss Please, take a look.

**Review by struss — APPROVED** [review]
LGTM! Sorry for late! Thanks!

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM
